Split `autodiff` into `autodiff_forward` and `autodiff_reverse`

Sa4dUs opened the pull request:
This PR splits `#[autodiff]` macro so `#[autodiff(df, Reverse, args)]` would become `#[autodiff_reverse(df, args)]` and `#[autodiff(df, Forward, args)]` would become `#[autodiff_forwad(df, args)]`.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

Sa4dUs committed:
Split `autodiff` into `autodiff_forward` and `autodiff_reverse`

Pending fix.
```
error: cannot find a built-in macro with name `autodiff_forward`
    --> library\core\src\macros\mod.rs:1542:5
     |
1542 | /     pub macro autodiff_forward($item:item) {
1543 | |         /* compiler built-in */
1544 | |     }
     | |_____^

error: cannot find a built-in macro with name `autodiff_reverse`
    --> library\core\src\macros\mod.rs:1549:5
     |
1549 | /     pub macro autodiff_reverse($item:item) {
1550 | |         /* compiler built-in */
1551 | |     }
     | |_____^

error: could not compile `core` (lib) due to 2 previous errors
```

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14853999947/job/41702974604) [(plain)](https://github.com/rust-lang/rust/commit/34f259b800c5bf4864be0f783029175cb26a9d63/checks/41702974604/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

jieyouxu commented:
I don't remember exactly, can built-in attributes be namespaced now? E.g. `#[autodiff::forward()]` and `#[autodiff::reverse()]`. That way, only one "top-level" name would cause macro ambiguities from the builtin attribute (`autodiff`), instead of two.

ZuseZ4 commented:
I think `::` looks much prettier than `_`. Our attribute is called `rustc_autodiff` and not autodiff, so I don't think there will be ambiguities either way.
@jdonszelmann is probably the right person to ask about parsing. (see the question above)

jieyouxu commented:
Right sorry, I keep forgetting there's two

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14915108868/job/41898750414) [(plain)](https://github.com/rust-lang/rust/commit/c47f0e36636c890bfa76c6579818aca423b21e16/checks/41898750414/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

Sa4dUs committed:
Disable autodiff bootstrapping

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14939791936/job/41974911350) [(plain)](https://github.com/rust-lang/rust/commit/3ab490be39a5ba91fcd5df196a1ef125a2d23c61/checks/41974911350/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 3.867 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.451 Collecting virtualenv
#12 4.494   Downloading virtualenv-20.31.2-py3-none-any.whl (6.1 MB)
#12 4.607      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 55.3 MB/s eta 0:00:00
#12 4.668 Collecting platformdirs<5,>=3.9.1
#12 4.673   Downloading platformdirs-4.3.8-py3-none-any.whl (18 kB)
#12 4.693 Collecting distlib<1,>=0.3.7
#12 4.698   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.706      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 79.2 MB/s eta 0:00:00
#12 4.745 Collecting filelock<4,>=3.12.2
#12 4.750   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.835 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 5.036 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.8 virtualenv-20.31.2
#12 5.036 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 5.2s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      147392 kB
DirectMap2M:     9289728 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.221
---
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6013 files
tidy check
tidy error: /checkout/compiler/rustc_builtin_macros/messages.ftl: message `builtin_macros_autodiff_mode` is not used
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/std/src/lib.rs:284: line not in alphabetical order
##[error]tidy error: /checkout/compiler/rustc_builtin_macros/src/lib.rs:12: line not in alphabetical order
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.1.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:46
  local time: Sat May 10 00:20:36 UTC 2025
  network time: Sat, 10 May 2025 00:20:36 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Sa4dUs committed:
Update codegen and pretty tests

UI tests are pending, will depend on error messages change.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14940052551/job/41975634518) [(plain)](https://github.com/rust-lang/rust/commit/b2cd498b591741c1b2d36793ed9275f2439f35a1/checks/41975634518/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.747 Building wheels for collected packages: reuse
#12 2.748   Building wheel for reuse (pyproject.toml): started
#12 2.962   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 2.963   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#12 2.963   Stored in directory: /tmp/pip-ephem-wheel-cache-d6g84x_s/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 2.965 Successfully built reuse
#12 2.966 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.355 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.356 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 3.898 Collecting virtualenv
#12 3.946   Downloading virtualenv-20.31.2-py3-none-any.whl (6.1 MB)
#12 4.017      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 90.8 MB/s eta 0:00:00
#12 4.075 Collecting platformdirs<5,>=3.9.1
#12 4.079   Downloading platformdirs-4.3.8-py3-none-any.whl (18 kB)
#12 4.097 Collecting distlib<1,>=0.3.7
#12 4.100   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.107      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 95.6 MB/s eta 0:00:00
#12 4.144 Collecting filelock<4,>=3.12.2
#12 4.148   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.230 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.421 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.8 virtualenv-20.31.2
#12 4.421 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.5s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      149440 kB
DirectMap2M:     8239104 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.218
---
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6013 files
tidy check
tidy error: /checkout/compiler/rustc_builtin_macros/messages.ftl: message `builtin_macros_autodiff_mode` is not used
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/std/src/lib.rs:284: line not in alphabetical order
##[error]tidy error: /checkout/compiler/rustc_builtin_macros/src/lib.rs:12: line not in alphabetical order
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.1.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:51
  local time: Sat May 10 00:48:54 UTC 2025
  network time: Sat, 10 May 2025 00:48:54 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Sa4dUs committed:
Initial naive implementation using `Symbols` to represent autodiff modes (`Forward`, `Reverse`)

Since the mode is no longer part of `meta_item`, we must insert it manually (otherwise macro expansion with `#[rustc_autodiff]` won't work).

This can be revised later if a more structured representation becomes necessary (using enums, annotated structs, etc).

Some tests are currently failing. I'll address them next.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14940248157/job/41976142181) [(plain)](https://github.com/rust-lang/rust/commit/02188404dfcc610e69d99f8bb57160ead8a74f8a/checks/41976142181/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.896 Building wheels for collected packages: reuse
#12 2.897   Building wheel for reuse (pyproject.toml): started
#12 3.108   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.109   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#12 3.109   Stored in directory: /tmp/pip-ephem-wheel-cache-2bhueouq/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.112 Successfully built reuse
#12 3.112 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.512 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.512 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.061 Collecting virtualenv
#12 4.119   Downloading virtualenv-20.31.2-py3-none-any.whl (6.1 MB)
#12 4.197      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 81.0 MB/s eta 0:00:00
#12 4.246 Collecting distlib<1,>=0.3.7
#12 4.254   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.260      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 96.8 MB/s eta 0:00:00
#12 4.298 Collecting platformdirs<5,>=3.9.1
#12 4.309   Downloading platformdirs-4.3.8-py3-none-any.whl (18 kB)
#12 4.350 Collecting filelock<4,>=3.12.2
#12 4.358   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.440 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.631 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.8 virtualenv-20.31.2
#12 4.632 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.7s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      133056 kB
DirectMap2M:     7206912 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.216
---
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6013 files
tidy check
tidy error: /checkout/compiler/rustc_builtin_macros/messages.ftl: message `builtin_macros_autodiff_mode` is not used
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.1.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:53
  local time: Sat May 10 01:10:54 UTC 2025
  network time: Sat, 10 May 2025 01:10:54 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

ZuseZ4 commented:
Try `build/x86_64-unknown-linux-gnu/llvm/bin/clang-format  compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp -i` (Adjusted for all the cpp files you touched).

Sa4dUs committed:
Update UI tests

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14949431593/job/41996945072) [(plain)](https://github.com/rust-lang/rust/commit/841ad9418d5fcf62bbf7dbb1da0edfb45901b8cd/checks/41996945072/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 20.6s done
#19 DONE 27.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
3    |
4 LL | use std::autodiff::autodiff;
-    |     ----------------------- previous import of the macro `autodiff` here
- ...
- LL | use my_macro::autodiff; // bring `autodiff` in scope
-    |     ^^^^^^^^^^^^^^^^^^ `autodiff` reimported here
+    |     ^^^^^^^^^^^^^^^^^^^^^^^ no `autodiff` in `autodiff`
9    |
-    = note: `autodiff` must be defined only once in the macro namespace of this module
- help: you can use `as` to change the binding name of the import
+ help: consider importing this module instead
12    |
- LL | use my_macro::autodiff as other_autodiff; // bring `autodiff` in scope
-    |                        +++++++++++++++++
+ LL - use std::autodiff::autodiff;
+ LL + use std::autodiff;
+    |
15 
- error: this rustc version does not support autodiff
-   --> $DIR/visibility.rs:14:1
+ error[E0432]: unresolved import `my_macro::autodiff`
+   --> $DIR/visibility.rs:12:5
18    |
- LL | #[autodiff]
-    | ^^^^^^^^^^^
+ LL | use my_macro::autodiff; // bring `autodiff` in scope
+    |     ^^^^^^^^^^^^^^^^^^ no `autodiff` in the root
+    |
+ help: consider importing this module instead
+    |
+ LL - use my_macro::autodiff; // bring `autodiff` in scope
+ LL + use std::autodiff; // bring `autodiff` in scope
+    |
21 
22 error: aborting due to 2 previous errors
23 

---
+ help: consider importing this module instead
+ LL - use std::autodiff::autodiff;
+ LL + use std::autodiff;
+    |
+ error[E0432]: unresolved import `my_macro::autodiff`
+   --> $DIR/visibility.rs:12:5
+ LL | use my_macro::autodiff; // bring `autodiff` in scope
+    |     ^^^^^^^^^^^^^^^^^^ no `autodiff` in the root
+    |
+ help: consider importing this module instead
+    |
+ LL - use my_macro::autodiff; // bring `autodiff` in scope
+ LL + use std::autodiff; // bring `autodiff` in scope
+    |
+ For more information about this error, try `rustc --explain E0432`.


The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args autodiff/visibility.rs`

error in revision `std_autodiff`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/autodiff/visibility.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "std_autodiff" "--check-cfg" "cfg(test,FALSE,std_autodiff,no_std_autodiff)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff/auxiliary" "--extern" "my_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff/auxiliary/libmy_macro.so"
stdout: none
--- stderr -------------------------------
error[E0432]: unresolved import `std::autodiff::autodiff`
##[error]  --> /checkout/tests/ui/autodiff/visibility.rs:9:5
   |
---
LL - use std::autodiff::autodiff;
LL + use std::autodiff;
   |

error[E0432]: unresolved import `my_macro::autodiff`
##[error]  --> /checkout/tests/ui/autodiff/visibility.rs:12:5
   |
LL | use my_macro::autodiff; // bring `autodiff` in scope
   |     ^^^^^^^^^^^^^^^^^^ no `autodiff` in the root
   |
help: consider importing this module instead
   |
LL - use my_macro::autodiff; // bring `autodiff` in scope
LL + use std::autodiff; // bring `autodiff` in scope
   |

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0432`.
------------------------------------------


---- [ui] tests/ui/autodiff/visibility.rs#no_std_autodiff stdout ----

error in revision `no_std_autodiff`: test compilation failed although it shouldn't!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/autodiff/visibility.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "no_std_autodiff" "--check-cfg" "cfg(test,FALSE,std_autodiff,no_std_autodiff)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.no_std_autodiff" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.no_std_autodiff/auxiliary" "--extern" "my_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.no_std_autodiff/auxiliary/libmy_macro.so"
stdout: none
--- stderr -------------------------------
error[E0432]: unresolved import `my_macro::autodiff`
##[error]  --> /checkout/tests/ui/autodiff/visibility.rs:12:5
   |
LL | use my_macro::autodiff; // bring `autodiff` in scope
   |     ^^^^^^^^^^^^^^^^^^ no `autodiff` in the root
   |
help: consider importing this module instead
   |
LL - use my_macro::autodiff; // bring `autodiff` in scope
LL + use std::autodiff; // bring `autodiff` in scope
   |

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0432`.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14949843653/job/41997797771) [(plain)](https://github.com/rust-lang/rust/commit/5450e7f359b7f6bc9e1571859079b67893a39daf/checks/41997797771/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 20.3s done
#18 DONE 34.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
3    |
4 LL | use std::autodiff::autodiff_forward;
-    |     ------------------------------- previous import of the macro `autodiff` here
- ...
+    |     ------------------------------- previous import of the macro `autodiff_forward` here
+ LL | extern crate my_macro;
7 LL | use my_macro::autodiff_forward; // bring `autodiff` in scope
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `autodiff` reimported here
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `autodiff_forward` reimported here
9    |
10    = note: `autodiff_forward` must be defined only once in the macro namespace of this module
11 help: you can use `as` to change the binding name of the import

12    |
- LL | use my_macro::autodiff_forward as other_autodiff_forward; // bring `autodiff_forward` in scope
-    |                                   ++++++++++++++++++++++
+ LL | use my_macro::autodiff_forward as other_autodiff_forward; // bring `autodiff` in scope
+    |                                +++++++++++++++++++++++++
15 
- error: this rustc version does not support autodiff_forward
-   --> $DIR/visibility.rs:14:1
+ error: this rustc version does not support autodiff
---
Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/autodiff/visibility.rs:11:5
-   --> /checkout/tests/ui/autodiff/visibility.rs:13:1
+   --> $DIR/visibility.rs:11:5
+    |     ------------------------------- previous import of the macro `autodiff_forward` here
+ LL | extern crate my_macro;
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `autodiff_forward` reimported here
+ LL | use my_macro::autodiff_forward as other_autodiff_forward; // bring `autodiff` in scope
+    |                                +++++++++++++++++++++++++
+ error: this rustc version does not support autodiff
+   --> $DIR/visibility.rs:13:1


The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args autodiff/visibility.rs`

error in revision `std_autodiff`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/autodiff/visibility.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "std_autodiff" "--check-cfg" "cfg(test,FALSE,std_autodiff,no_std_autodiff)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff/auxiliary" "--extern" "my_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff/auxiliary/libmy_macro.so"
stdout: none
--- stderr -------------------------------
error[E0252]: the name `autodiff_forward` is defined multiple times
##[error]  --> /checkout/tests/ui/autodiff/visibility.rs:11:5
   |
LL | use std::autodiff::autodiff_forward;
   |     ------------------------------- previous import of the macro `autodiff_forward` here
LL | extern crate my_macro;
LL | use my_macro::autodiff_forward; // bring `autodiff` in scope
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `autodiff_forward` reimported here
   |
   = note: `autodiff_forward` must be defined only once in the macro namespace of this module
help: you can use `as` to change the binding name of the import
   |
LL | use my_macro::autodiff_forward as other_autodiff_forward; // bring `autodiff` in scope
   |                                +++++++++++++++++++++++++

error: this rustc version does not support autodiff
##[error]  --> /checkout/tests/ui/autodiff/visibility.rs:13:1
   |
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14950086330/job/41998308758) [(plain)](https://github.com/rust-lang/rust/commit/375005ac73478fc3ab64948260decc3e9b42dc67/checks/41998308758/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 20.6s done
#18 DONE 28.6s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
3    |
4 LL | use std::autodiff::autodiff_forward;
5    |     ------------------------------- previous import of the macro `autodiff_forward` here
- ...
+ LL | extern crate my_macro;
7 LL | use my_macro::autodiff_forward; // bring `autodiff_forward` in scope
8    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `autodiff_forward` reimported here
9    |

13 LL | use my_macro::autodiff_forward as other_autodiff_forward; // bring `autodiff_forward` in scope
14    |                                +++++++++++++++++++++++++
15 
- error: this rustc version does not support autodiff_forward
+ error: this rustc version does not support autodiff
17   --> $DIR/visibility.rs:13:1
---
To only update this specific test, also pass `--test-args autodiff/visibility.rs`

error in revision `std_autodiff`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/autodiff/visibility.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "std_autodiff" "--check-cfg" "cfg(test,FALSE,std_autodiff,no_std_autodiff)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff/auxiliary" "--extern" "my_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff/auxiliary/libmy_macro.so"
stdout: none
--- stderr -------------------------------
error[E0252]: the name `autodiff_forward` is defined multiple times
##[error]  --> /checkout/tests/ui/autodiff/visibility.rs:11:5
   |
LL | use std::autodiff::autodiff_forward;
   |     ------------------------------- previous import of the macro `autodiff_forward` here
LL | extern crate my_macro;
LL | use my_macro::autodiff_forward; // bring `autodiff_forward` in scope
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `autodiff_forward` reimported here
   |
   = note: `autodiff_forward` must be defined only once in the macro namespace of this module
help: you can use `as` to change the binding name of the import
   |
LL | use my_macro::autodiff_forward as other_autodiff_forward; // bring `autodiff_forward` in scope
   |                                +++++++++++++++++++++++++

error: this rustc version does not support autodiff
##[error]  --> /checkout/tests/ui/autodiff/visibility.rs:13:1
   |
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14954582343/job/42008551434) [(plain)](https://github.com/rust-lang/rust/commit/71d0e7c659963b3625880743c7d6311ada5b04f1/checks/42008551434/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 20.6s done
#18 DONE 28.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/autodiff/visibility.rs#std_autodiff stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff/visibility.std_autodiff.stderr"
diff of stderr:

13 LL | use my_macro::autodiff_forward as other_autodiff_forward; // bring `autodiff_forward` in scope
14    |                                +++++++++++++++++++++++++
15 
- error: aborting due to 1 previous error
+ error: this rustc version does not support autodiff
+   --> $DIR/visibility.rs:13:1
+    |
+ LL | #[autodiff_forward(dfoo)]
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^
+ 
+ error: aborting due to 2 previous errors
17 
18 For more information about this error, try `rustc --explain E0252`.
19 

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/autodiff/visibility.rs:13:1
+ error: this rustc version does not support autodiff
+   --> $DIR/visibility.rs:13:1
+    |
+ LL | #[autodiff_forward(dfoo)]
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^
+ 
+ error: aborting due to 2 previous errors


The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args autodiff/visibility.rs`

error in revision `std_autodiff`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/autodiff/visibility.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "std_autodiff" "--check-cfg" "cfg(test,FALSE,std_autodiff,no_std_autodiff)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff/auxiliary" "--extern" "my_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/autodiff/visibility.std_autodiff/auxiliary/libmy_macro.so"
stdout: none
--- stderr -------------------------------
error[E0252]: the name `autodiff_forward` is defined multiple times
##[error]  --> /checkout/tests/ui/autodiff/visibility.rs:11:5
   |
LL | use std::autodiff::autodiff_forward;
   |     ------------------------------- previous import of the macro `autodiff_forward` here
LL | extern crate my_macro;
LL | use my_macro::autodiff_forward; // bring `autodiff_forward` in scope
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `autodiff_forward` reimported here
   |
   = note: `autodiff_forward` must be defined only once in the macro namespace of this module
help: you can use `as` to change the binding name of the import
   |
LL | use my_macro::autodiff_forward as other_autodiff_forward; // bring `autodiff_forward` in scope
   |                                +++++++++++++++++++++++++

error: this rustc version does not support autodiff
##[error]  --> /checkout/tests/ui/autodiff/visibility.rs:13:1
   |
LL | #[autodiff_forward(dfoo)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0252`.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14955195604/job/42009915508) [(plain)](https://github.com/rust-lang/rust/commit/aa6d358eec37d18215ac0f57cc2844b9a299dfe1/checks/42009915508/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 22.0s done
#19 DONE 27.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
[RUSTC-TIMING] proc_macro test:false 5.886
[RUSTC-TIMING] test test:false 9.406
    Finished `release` profile [optimized] target(s) in 1m 20s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 80.838
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.804
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 3.019
    Finished `release` profile [optimized] target(s) in 4.16s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.210
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.183
---
---- library/core/src/macros/mod.rs - macros::builtin::autodiff_forward (line 1535) stdout ----
error: cannot find attribute `autodiff_forward` in this scope
 --> library/core/src/macros/mod.rs:1536:3
  |
4 | #[autodiff_forward(df, Dual, Dual, Dual)]
  |   ^^^^^^^^^^^^^^^^
  |
help: consider importing this attribute macro
  |
3 + use std::autodiff::autodiff_forward;
---
---- library/core/src/macros/mod.rs - macros::builtin::autodiff_reverse (line 1563) stdout ----
error: cannot find attribute `autodiff_reverse` in this scope
 --> library/core/src/macros/mod.rs:1564:3
  |
4 | #[autodiff_reverse(df, Active, Active, Active)]
  |   ^^^^^^^^^^^^^^^^
  |
help: consider importing this attribute macro
  |
3 + use std::autodiff::autodiff_reverse;
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14955893549/job/42011490633) [(plain)](https://github.com/rust-lang/rust/commit/6d3019f5ba07236d653006f47ee215a05917f08b/checks/42011490633/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 19.9s done
#18 DONE 27.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
[RUSTC-TIMING] proc_macro test:false 5.532
[RUSTC-TIMING] test test:false 9.512
    Finished `release` profile [optimized] target(s) in 1m 19s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 79.384
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.780
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 2.896
    Finished `release` profile [optimized] target(s) in 4.09s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.203
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.183
---

error: could not document `core`
warning: build failed, waiting for other jobs to finish...
[RUSTC-TIMING] core test:false 27.701
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:37:03
  local time: Sun May 11 13:21:48 UTC 2025
  network time: Sun, 11 May 2025 13:21:49 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Sa4dUs committed:
Update generic tests

ZuseZ4 commented on compiler/rustc_builtin_macros/src/lib.rs:
```diff
@@ -5,10 +5,11 @@
 #![allow(internal_features)]
 #![allow(rustc::diagnostic_outside_of_impl)]
 #![allow(rustc::untranslatable_diagnostic)]
+#![cfg_attr(bootstrap, feature(let_chains))]
```
this seems unrelated?

Sa4dUs replied:
oops, must’ve slipped in while resolving merge conflicts, fixed

ZuseZ4 commented:
Now that we can't get the Forward/Reverse symbols from the input, using sym to create them seems like the right thing to do. We already have a few UI tests to make sure that our current autodiff symbol there doesn't colide with similar named user types, so adding Forward/Reverse to the list should also be fine.
lgtm, with the small doc improvement.

Sa4dUs committed:
Minor macro docs fixes

Sa4dUs commented on library/core/src/macros/mod.rs:
```diff
@@ -1519,20 +1519,40 @@ pub(crate) mod builtin {
     }
 
+    /// the derivative of a given function in the forward mode of differentiation.
+    /// It may only be applied to a function.
```
Okey, done.

ZuseZ4 commented:
@bors r+

bors commented:
:pushpin: Commit c6c2fde737e5d6730cc7135afd841a0afed6096e has been approved by `ZuseZ4`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ZuseZ4 c6c2fde737e5d6730cc7135afd841a0afed6096e -->
<!-- homu: {"type":"Approved","sha":"c6c2fde737e5d6730cc7135afd841a0afed6096e","approver":"ZuseZ4","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
